File: 05_fp_challenges/06_daily_temperatures/Solution06.hs
-- Daily Temperatures  (LeetCode 739)
-- ===================================
-- For each day, how many days until a WARMER temperature? 0 if none.
--   dailyTemperatures [73,74,75,71,69,72,76,73] == [1,1,4,2,1,1,0,0]
--   dailyTemperatures [30,40,50,60] == [1,1,1,0]
--   dailyTemperatures [30,60,90] == [1,1,0]

module Main where

popping :: [(Int, Int)] -> Int -> [(Int, Int)]
popping [] _ = []
popping total@(s : stack) num = if snd s > num then total else popping stack num

dailyTemperatures :: [Int] -> [Int]
dailyTemperatures temps =
  fst
    ( foldr
        ( \(temp, idx) (ans, stack) ->
            ( fst
                ( head
                    ( case popping stack temp of
                        [] -> (idx, 0) : []
                        other -> other
                    )
                )
                - idx
                : ans,
              ((idx, temp) : popping stack temp)
            )
        )
        ([], [])
        (zip temps [0 ..])
    )

-- ---- tests ----
main :: IO ()
main = do
  check (dailyTemperatures [73, 74, 75, 71, 69, 72, 76, 73]) [1, 1, 4, 2, 1, 1, 0, 0]
  check (dailyTemperatures [30, 40, 50, 60]) [1, 1, 1, 0]
  check (dailyTemperatures [30, 60, 90]) [1, 1, 0]
  putStrLn "all passed"
  where
    check got want
      | got == want = putStrLn ("  PASS  " ++ show got)
      | otherwise = error ("  FAIL  got " ++ show got ++ " want " ++ show want)

File: 05_fp_challenges/06_daily_temperatures/Solution06.java
// Daily Temperatures  (LeetCode 739)
// ===================================
// For each day, how many days until a WARMER temperature? 0 if none.
//   dailyTemperatures([73,74,75,71,69,72,76,73]) == [1,1,4,2,1,1,0,0]
//   dailyTemperatures([30,40,50,60]) == [1,1,1,0]

import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;

public class Solution06 {

    static int[] dailyTemperatures(int[] temps) {
        throw new UnsupportedOperationException("TODO");
    }

    // ---- tests ----
    public static void main(String[] args) {
        check(dailyTemperatures(new int[]{73,74,75,71,69,72,76,73}), new int[]{1,1,4,2,1,1,0,0});
        check(dailyTemperatures(new int[]{30,40,50,60}),             new int[]{1,1,1,0});
        check(dailyTemperatures(new int[]{30,60,90}),                new int[]{1,1,0});
        System.out.println("all passed");
    }

    static void check(int[] got, int[] want) {
        if (!Arrays.equals(got, want))
            throw new AssertionError("FAIL got " + Arrays.toString(got) + " want " + Arrays.toString(want));
        System.out.println("  PASS  " + Arrays.toString(got));
    }
}
